### PR TITLE
feat: edit the bug report before submission

### LINKED ISSUE #3757: Privacy concerns about automatic starship bug-report
#### Is your feature request related to a problem? Please describe.

When I ran `starship bug-report` my private shell configuration scripts were submitted without my review and consent through a long URL. :-( Luckily the NoScript browser extension blocked that request for me.

#### Describe the solution you'd like

Open the auto-generated report with `$EDITOR` and let me edit it before it is sent.

#### Describe alternatives you've considered

Better also provide a command line flag to print the report to terminal.

### PR DESCRIPTION
**Pull request by IsaacHorvath**
#### Description
This pull request changes the behaviour of `starship bug-report` to drop into the system editor and allow the user to edit the bug report before populating the GitHub post. An empty bug report cancels the post. I've also added a `-p` `--print` flag to dump the report to stdin instead of posting, as requested in the issue.

~Note: This PR adds a crate: [`edit`](https://crates.io/crates/edit).~

#### Motivation and Context
To allow for better privacy and control over data submitted in bug reports.
Closes #3757 

#### How Has This Been Tested?

- [ ] I have tested using **MacOS**
- [x] I have tested using **Linux**
- [ ] I have tested using **Windows**

#### Checklist:

- [x] I have updated the tests accordingly.
- [ ] I have updated the documentation accordingly. 
_I can't find any documentation for `starship bug-report`_

### CODE REVIEW

**Comment by davidkna** [comment]
You can probably use `open` instead of `edit`

**Comment by IsaacHorvath** [comment]
I used `edit` because it automatically uses a system temporary file, has more robust means of finding a system editor, and the code is a lot more concise - this block handles the whole edit process:
```
    let report = match edit::edit(report) {
        Ok(r) => r,
        Err(e) => {
            println!("Error editing bug report: {}", e);
            return;
        }
    };
```
However, if you want to keep a tighter supply chain, I'm happy to abstract out the code in the configure module and rebuild using that. Alternatively, I could clean up `configure.rs` and have it use `edit` as well for simplicity's sake.

**Comment by davidkna** [comment]
`edit` looks nice, but it doesn't seem to properly handle shell quoting, and it doesn't seem to be very popular.

File: src/utils.rs [hunk 1]
```diff
@@ -96,6 +102,65 @@ impl PartialEq for CommandOutput {
...
+    let editor_cmd = shell_words::split(&get_editor()).expect("Unmatched quotes found in $EDITOR.");
+
+    let command = create_command(&editor_cmd[0])
+        .expect("Unable to locate editor in $PATH.")
+        .stdin(Stdio::inherit())
+        .stdout(Stdio::inherit())
+        .stderr(Stdio::inherit())
+        .args(&editor_cmd[1..])
+        .arg(path)
+        .status();
```
Comment (davidkna) [inline src/utils.rs:130]: I'd prefer improving error handling on editing before going forward with this change (https://github.com/starship/starship/issues/3758 #3766)

File: src/bug_report.rs [hunk 2]
```diff
@@ -17,7 +17,27 @@ pub fn create() {
...
+        Ok(r) => r,
+        Err(e) => {
+            eprintln!("Error editing bug report: {}", e);
+            return;
+        }
+    };
+
+    if report.is_empty() {
+        return;
+    };
```
Comment (davidkna) [inline src/bug_report.rs:38]: Please print a message before exiting, or preferably just accept empty reports?

File: src/bug_report.rs [hunk 3]
```diff
@@ -17,7 +17,27 @@ pub fn create() {
...
+    }
+
+    // Drop to system editor in a temp file
+    let report = match utils::edit_temp(report) {
+        Ok(r) => r,
+        Err(e) => {
+            eprintln!("Error editing bug report: {}", e);
+            return;
+        }
+    };
```
Comment (davidkna) [inline src/bug_report.rs:34]: Perhaps ask for confirmation before sending? Some editors might not wait for exit like `code` without `--wait`.

A flag for skipping editing in general might be good too (and reference it in relevant error messages).

File: src/utils.rs [hunk 4]
```diff
@@ -96,6 +102,65 @@ impl PartialEq for CommandOutput {
...
+    let fp = file.as_os_str().to_os_string();
+    edit(fp);
+    let buffer = read_file(&file)?;
+
+    file.close()?;
+    Ok(buffer)
+}
+
+/// Open the given path in the system editor
+pub fn edit(path: OsString) {
```
Comment (davidkna) [inline src/utils.rs:120]: Why not `&Path`?

**Review by davidkna — CHANGES_REQUESTED** [review]

**Comment by davidkna** [comment]
Please also rebase this PR to add the codecov check.

**Pull request closed without merging**